Changes for pfte

rpiazza opened the pull request:
Can you take a look at this to see if I'm on the right track??

rpiazza committed:
Bump version: 0.1.0 → 0.2.0

rpiazza committed:
maybe oneOf with weights

chisholm commented on stix2generator/stix21_registry.json:
```diff
@@ -2771,15 +2778,15 @@
+            "10": "malicious-activity",
+            "10": "attribution",
+            "40": "unknown"
+        }
```
This JSON object has duplicate keys, so it wouldn't work.  Weights need not be distinct, therefore we can't use them as keys.  For full generality, the things being weighted would need to be arbitrary specs (not just strings), so they can't be keys either.  If the value of "oneOf" were a list of two-element [weight, spec] lists, it would be ambiguous since all non-object JSON values already have the literal interpretation as a spec which generates that fixed value.  E.g. does a [weight,spec] value mean to generate that fixed list, or something else?  I think an array value for "oneOf" will need to retain its current semantics.  The only way I can think to have an unambiguous alternative structure is to add support for an object value here.  E.g. {"weights": [...], "specs": [...]}.

But there are currently no probabilities anywhere in the spec structure, and I kind of like the idea of keeping it more declarative and not adding behavioral complexity like specifying probabilities.  Different users might want to use different probabilities, so it doesn't seem like the kind of thing which should be baked into the specifications.  Or it should at least be overridable somehow.  One could also imagine defining probabilities and/or distributions over more than just "oneOf" specs.  One could define probabilities for optional properties of objects, and distributions for boolean specs (e.g. probability of true vs false), maybe numeric ranges, ...  A distribution doesn't make sense for optional properties, since there is no need to pick any of them.  In fact, the "minimize_ref_properties" config setting could be re-conceptualized as a shortcut for setting p=0 for all ref properties.  A distribution is necessary for something like a boolean spec, since either a true or false value must *always* be generated in that case.  Turning down the probability for one outcome must necessarily push the probability up for other outcomes.

Anyway, providing probabilities separately from the registry, e.g. as a config item, would require the ability to specify what to apply the probabilities to.  E.g. to either refer to a component of a specification, or a component of a JSON value being generated.  Maybe a JSONPath-like idea?  Hmmmm.

rpiazza replied:
> This JSON object has duplicate keys, so it wouldn't work. 

Of course, I was trying to do it the "other way" - i.e., the probability being the value, but that didn't work when I tried it for indicator/valid_until - as you also say.  I might try the "unambiguous alternative structure" you suggest.  If it works, we can move on to your hesitancy about putting probabilities in the spec structure.

I was thinking that if different users wanted different probabilities, we could use the override capability that already exists.  We could add an argument to the stix-generator command line to take in the override file.  

I understand this is not a complete solution for probabilistic choices - but maybe a stop-gap solution?

I'm open to a config item - if you want to explore it.

rpiazza committed:
change format for weighted choices spec

rpiazza commented:
See if this is more pleasing.

rpiazza committed:
revert stix21-registry - changes moved to pfte

rpiazza commented:
Can you review this to see if you think the latest is good functionality to add to the generator?

If you do, I write some doc for it, and we can release a new patch or minor version

rpiazza commented:
see ais_registry.json in pftegen for an example of the probabilistic choice spec.

chisholm commented:
Github is not letting me add inline code comments... it seems to be behaving as if I'm not logged in.  So I'll have to comment here.

At:

https://github.com/oasis-open/cti-stix-generator/blob/c7337312da41db051b034f974380056cf1088648/stix2generator/__init__.py#L133

There are two issues:
- The code in that block was written generically, with `if isinstance(val, Mapping):`, which admits more than dicts as mapping types.  So within that if statement, anything specific to dicts should be avoided.
- An empty dict should not be created just to check whether your dict is empty.  Empty dicts are falsey, and let's assume that's true for whatever the mapping type is too.  So just use `if not val:`.

I think code should be added to ensure all weights are non-negative.  I tried using `random.choices()` with negative weights and there was no error, but I got funny results.  Since the `random` library code doesn't produce an error, we should produce an error.  Zeros are apparently okay and just cause a choice to never be chosen.

I like that object-valued "oneOf" is not used in the builtin object generator registry, but we still ought to have a unit test for it.  Maybe add to `test_object_generator_misc.py`.

rpiazza commented:
Is there some checking done on the registry up-front, or does it have to be done in __generate_plain.

It probably makes sense to check to see if the weights add up to 100.  

Any ideas how we should handle negative or < 100 issues?

rpiazza commented:
Any suggestions on how to structure the test, since the generated thing is probabilistic?

chisholm commented:
The weights don't have to add to 100, nor do they have to be integers.  They can be any non-negative real number.  Actually, at least one weight should be positive too.  I tried using all zeros, and that also didn't produce an error, but also produced funny results.  I'm just trying to avoid unexpected behavior whose cause is hard for people to track down.  So I was thinking we just raise an exception.

There are no registry sanity checks, but the specs used for generation don't necessarily come from the registry anyway.  You could call generate_from_spec() and generate a value based on any spec you pass in.  A registry is convenient, but optional.

At minimum, a unit test could include runs with that spec structure, just to ensure there are no errors.  You could have some tests which force some testable predictability, e.g. make one of the weights zero and ensure that value never occurs.  But yeah, if output is unpredictable, there's not much you can do to check it.

rpiazza commented:
what do you get if the numbers don't add up to 100 - is it just normalized?  That's what the documentation seems to say.
It's kind weird - but I guess there is no natural way to specify the weights.

First you say: They can be any non-negative real number
Then you say:  Actually, at least one weight should be positive too

so can they be negative or not?

chisholm commented:
The second statement further qualifies the first.  There are no inconsistencies that I can see.  Yes, the weights are normalized to obtain a probability distribution.  I think it's a pretty standard behavior; it means relative probabilities are easy to express (e.g. one value should be "twice as likely" as another), and you don't have to worry about getting them to add up right.

rpiazza committed:
add test for prop choices

rpiazza committed:
handle style, etc

rpiazza commented:
@chisholm - if you can look at the test I added to see if it makes sense, I think we can merge this in...

chisholm committed:
Fix mac-style eol mangling on test_object_generator_misc.py

chisholm committed:
Add some sanity checks on weights for object-valued "oneOf"
specs.  Add some unit tests for those sanity checks.  Also
change the weighted oneOf unit test so that it's guaranteed
to pass, and so that the repetition repeats a complete test
(it should not *be* the test).

chisholm committed:
pre-commit style fixes

chisholm committed:
Add a unit test for when there are no weights (should cause an
error).  Also, adjust the weight sanity checks to only trigger
the zero weights error if any weights were given.  It doesn't
really make any sense to say "Weights may not all be zero" if
there were no weights at all.

chisholm committed:
Add some documentation for the alternate oneOf structure of
object generator specifications.

chisholm committed:
Remove a reference to the 'unicode' Python type in the object
generator documentation.  It's a python 2 type, and the library
no longer supports python 2.
